### PR TITLE
Add blog post: Using Finalizers to Control Deletion

### PR DESCRIPTION
**Pull request by aaron-kasten**
Add block post describing operation of finalizers and owner references via `kubectl delete`

This blog post contains the content of a presentation given at KubeCon, in blog form.

Submission to `kubenetes.io` recommended by CNCF.

### CODE REVIEW

**Comment by aaron-kasten** [comment]
@onlydole @sftim email address on git commit is that of our corporate CLA (`aaron@kasten.io` for a CLA for `kasten.io`).

I'm not sure what's missing from the PR to get the CLA detection working correctly.

File: content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+$ kubectl delete configmap/mymap
+configmap "mymap" deleted
+```
+
+```
+$ kubectl get configmap/mymap
+Error from server (NotFound): configmaps "mymap" not found
+```
+
+All of the shell commands are in **bold**, and the output follows. You can see that we begin with a `kubectl create configmap mymap`, which will create the empty configmap `mymap`. Next, we need to `get` the configmap to prove it exists. We can then delete that configmap. Attempting to `get` it again produces an HTTP 404 error, which means the configmap is not found.
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:48]: The shell commands are not in bold.

File: content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+## Key Takeaways
+
+As these examples demonstrate, finalizers can get in the way of deleting resources in Kubernetes, especially when there are parent-child relationships between objects. Often, there is a reason for adding a finalizer into the code, so you should always investigate before manually deleting it. Owner references allow you to specify and remove trees of resources, although finalizers will be honored in the process. Finally, the propagation policy can be used to specify the order of deletion via a custom API call, giving you control over how objects are deleted. Now that you know a little more about how deletions work in Kubernetes, we recommend you try it out on your own, using a test cluster. 
+
+All of the scenarios covered in this post are demonstrated in a recent webinar; you can [watch the recording here](https://www.youtube.com/watch?v=F7-ZxWwf4sY).
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:269]: If you want you can embed the video in the page, too. See https://gohugo.io/content-management/shortcodes/#youtube

File: content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+- `Foreground`: Children are deleted before the parent (post-order)
+- `Background`: Parent is deleted before the children (pre-order)
+- `Orphan`: Owner references are ignored
+
+Keep in mind that when you delete an object and owner references have been specified, finalizers will be honored in the process. This can result in trees of objects persisting, and you end up with a partial deletion. At that point, you have to look at any existing owner references on your objects, as well as any finalizers, to understand what’s happening. 
+
+## Forcing a Deletion of a Namespace
+
+There's one situation that may require forcing finalization for a namespace. If you've deleted a namespace and you've cleaned out all of the objects under it, but the namespace still exists, use the API call `finalize`.This informs the namespace controller that it needs to remove the finalizer from the namespace and perform any cleanup:
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:243]: Is `finalize` an API call or a resource? It looks like a resource (you can `PUT` it) but the resource name is a verb.

File: content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+title: 'Deletes with Finaliers and Owner References'
+date: 2021-03-03
+slug: deletes-finaliers-and-owner-references
+---
+
+**Authors:** Aaron Alpar (Kasten)
+
+Deleting objects in Kubernetes can be challenging. You may think you’ve deleted something, only to find it still persists. While issuing a `kubectl delete` command and hoping for the best might work for day-to-day operations, understanding how Kubernetes `delete` commands operate will help you understand why some objects linger after deletion. 
+
+In this post, we’ll look at:
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md:12]: In docs (and documentation-like blog articles), [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we):
```suggestion
In this post, I’ll look at:
```

**Review by sftim — COMMENTED** [review]
Hi. “Finaliers” is misspelled, it should be Finalizers.

I've added some other feedback.

Would you be interested in turning this page into a task? For example, “Use finalizers to control deletion” could be an appropriate title.

File: content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md (hunk 3 above)
Comment (aaron-kasten) [reply content/en/blog/_posts/2021-03-03-deletes-finaliers-and-owner-references.md]: Changed - hopefully clearer.

Here's a citation to the technique: https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-408599873

**Comment by aaron-kasten** [comment]
@sftim thank you for the feedback!  I integrated the changes - hopefully an improvement.

**Comment by aaron-kasten** [comment]
@sftim what is the next step for approval?  Are there any actions I need to take on this?

File: content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+## Key Takeaways
+
+As these examples demonstrate, finalizers can get in the way of deleting resources in Kubernetes, especially when there are parent-child relationships between objects. Often, there is a reason for adding a finalizer into the code, so you should always investigate before manually deleting it. Owner references allow you to specify and remove trees of resources, although finalizers will be honored in the process. Finally, the propagation policy can be used to specify the order of deletion via a custom API call, giving you control over how objects are deleted. Now that you know a little more about how deletions work in Kubernetes, we recommend you try it out on your own, using a test cluster. 
+
+All of the scenarios covered in this post are demonstrated in this video from a recent webinar: {{< youtube id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:269]: For better sizing, try:
```suggestion
{{< youtube youtube-quote-sm id="F7-ZxWwf4sY" title="Clean Up Your Room! What Does It Mean to Delete Something in K8s">}}
```

File: content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+```
+$ kubectl get configmap/mymap
+Error from server (NotFound): configmaps "mymap" not found
+```
+
+Shell commands preceded by `$` are followed by their output. You can see that we begin with a `kubectl create configmap mymap`, which will create the empty configmap `mymap`. Next, we need to `get` the configmap to prove it exists. We can then delete that configmap. Attempting to `get` it again produces an HTTP 404 error, which means the configmap is not found.
+
+The state diagram for the basic `delete` command is very simple:
+
+![State diagram for delete](/images/blog/2021-03-03-use-finalizers-to-control-deletion/state-diagram-delete.png)
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:52]: This doesn't render - see https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/03/03/using-finalizers-to-control-deletion/

**Review by sftim — CHANGES_REQUESTED** [review]
A couple more fixes needed I think. Also, would you be willing to squash the commits in this article @aaron-kasten 

(if you're not sure but you're happy to have the changes combined, we can do that squashing for you).

File: content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md (hunk 6 above)
Comment (aaron-kasten) [reply content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:52]: interesting ... I'll reproduce and fix here.

**Comment by aaron-kasten** [comment]
@sftim:
* fixed image references (I'll know better for checking image references next time).
* shrunk images a little
* squashed local branch (appears in PR as one commit)
* merged your changes for Youtube link prologue (now gone)

I think that's it... tell me if there's anything else.

**Comment by aaron-kasten** [comment]
@sftim how does this PR look.  Is it ready for merge?

**Comment by sftim** [comment]
/retitle Add blog post: Using Finalizers to Control Deletion

**Comment by sftim** [comment]
Preview: [Using Finalizers to Control Deletion](https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/03/03/using-finalizers-to-control-deletion/)

**Comment by sftim** [comment]
We'll follow the [review process](https://github.com/kubernetes/community/tree/master/sig-docs/blog-subproject#review-process) or something roughly like it. At the moment the blog team is a little short on volunteers I'm afraid.

**Comment by aaron-kasten** [comment]
@sftim I understand.  Tell me if there's anything I can do to help it along.

File: content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
+---
+layout: blog
+title: 'Using Finalizers to Control Deletion'
+date: 2021-03-03
```
Comment (sftim) [inline content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md:4]: Can I pencil in a future date here?
```suggestion
date: 2021-04-28
```

Comment (aaron-kasten) [reply content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md]: 2021-04-28 works.

Comment (aaron-kasten) [reply content/en/blog/_posts/2021-03-03-using-finalizers-to-control-deletion.md]: I used your suggested date.

**Review by sftim — COMMENTED** [review]
If you have time, it's tidier to make the filename match the publication date.

I hope we'll get round to a full review soon(ish).

**Comment by aaron-kasten** [comment]
> If you have time, it's tidier to make the filename match the publication date.
> 
> I hope we'll get round to a full review soon(ish).

Ugh ... didn't see that - apologies, I'm moving too fast.   I'll change and update.

**Comment by aaron-kasten** [comment]
> If you have time, it's tidier to make the filename match the publication date.
> 
> I hope we'll get round to a full review soon(ish).

Looks good now.  Opening for review ...

**Comment by kbhawkey** [comment]
/remove-language zh ja ko fr es pt vi id ru pl

**Comment by sftim** [comment]
I'll try to review this this week.

File: content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+Deleting objects in Kubernetes can be challenging. You may think you’ve deleted something, only to find it still persists. While issuing a `kubectl delete` command and hoping for the best might work for day-to-day operations, understanding how Kubernetes `delete` commands operate will help you understand why some objects linger after deletion. 
+
+In this post, I’ll look at:
+
+- What properties of a resource govern deletion
+- How finalizers and owner references impact object deletion
+- How the propagation policy can be used to change the order of deletions
+- How deletion works, with examples
+
+For simplicity, all examples will use configmaps and basic shell commands to demonstrate the process. We’ll explore how the commands work and discuss repercussions and results from using them in practice.
```
Comment (sftim) [inline content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:19]: nit:
```suggestion
For simplicity, all examples will use ConfigMaps and basic shell commands to demonstrate the process. We’ll explore how the commands work and discuss repercussions and results from using them in practice.
```

**Review by sftim — COMMENTED** [review]
/lgtm

File: content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md (hunk 8 above)
Comment (aaron-kasten) [reply content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:19]: lets go with that!

**Comment by aaron-kasten** [comment]
@sftim looks like the suggested change removed the `lgtm` label... I think it needs another `lgtm` (post change).

File: content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+  name: mymap
+  namespace: default
+  resourceVersion: "311456"
+  selfLink: /api/v1/namespaces/default/configmaps/mymap
+  uid: 93a37fed-23e3-45e8-b6ee-b2521db81638
+```
+
+In short, what’s happened is that the object was updated, not deleted. That’s because Kubernetes saw that the object contained finalizers and put it into a read-only state. The deletion timestamp signals that the object can only be read, with the exception of removing the finalizer key updates. In other words, the deletion will not be complete until we edit the object and remove the finalizer.
+
+Here's a demonstration of using the `patch` command to remove finalizers. If we want to delete and object, we can simply patch it on the command line to remove the finalizers. In this way, the deletion that was running in the background will complete and the object will be deleted. When we attempt to `get` that configmap, it will be gone.
```
Comment (rolfedh) [inline content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:114]: s/and object/an object/

File: content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+$ kubectl delete --cascade=false configmap/mymap-parent
+configmap "mymap-parent" deleted
+
+$ kubectl get configmap
+NAME          DATA   AGE
+mymap-child   0      13m21s
+```
+
+The --cascade option links to the propagation policy in the API, which allows you to change the order in which objects are deleted within a tree. In the following example uses API access to craft a custom delete API call with a the background propagation policy:
```
Comment (rolfedh) [inline content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:210]: s/a the/the/

**Comment by aaron-kasten** [comment]
@rolfedh fixups applied.   thank you.

**Review by rolfedh — APPROVED** [review]

**Review by sftim — COMMENTED** [review]
If you can bump the article to say Friday, 2021-04-30, @aaron-kasten

then I think we can probably get this article over the line for that.

Please change both the filename and the `date` in the front matter.

**Comment by nate-double-u** [comment]
Hi, I've gone through and tested all the examples in this post verbatim, and they all worked as I expected them to, until I got to the _Forcing a Deletion of a Namespace_ section.

This may be my own inexperience with k8s, but I'm not sure how to set up this example's environment, or see that the snippet did what I was expecting it to do.

Sorry for the long log dump, but I figured it was the best way to show what I saw. I'm using `minikube`, and was trying to correlate outcomes between the CI and the Kubernetes Dashboard.

I had run `kubectl proxy --port=8080 &` earlier in the process 
```
[1] 2817582
Starting to serve on 127.0.0.1:8080
```

Example log leading up to the _Forcing a Deletion of a Namespace_ section:
```
➜  blog-test cat <<EOF | kubectl create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: mymap-parent
EOF
configmap/mymap-parent created

➜  blog-test CM_UID=$(kubectl get configmap mymap-parent -o jsonpath="{.metadata.uid}")

➜  blog-test cat <<EOF | kubectl create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: mymap-child
  ownerReferences:
  - apiVersion: v1
    kind: ConfigMap
    name: mymap-parent
    uid: $CM_UID
EOF
configmap/mymap-child created

➜  blog-test cat <<EOF | curl -X PUT \
  localhost:8080/api/v1/namespaces/test/finalize \
  -H "Content-Type: application/json" \
  --data-binary @-
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "finalizers": null,
  }
}
EOF
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}%                                                                                                                                                       

➜  blog-test curl -X DELETE \
localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{
"kind":"DeleteOptions",
"apiVersion":"v1",
"propagationPolicy":"Background"
}' \
  -H "Content-Type: application/json"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Success",
  "details": { ... }
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Success",
  "details": {
    "name": "mymap-parent",
    "kind": "configmaps",
    "uid": "caad82b7-3308-479e-8e8d-b322a25f565b"
  }
}kind:: command not found
apiVersion:: command not found
metadata:: command not found
status:: command not found
details:: command not found

➜  blog-test cat <<EOF | curl -X PUT \
  localhost:8080/api/v1/namespaces/test/finalize \
  -H "Content-Type: application/json" \
  --data-binary @-
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "finalizers": null,
  }
}
EOF
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}% 
```

**Comment by nate-double-u** [comment]
(sorry, heads up to the GitHub email notification system users, I've edited the log in my previous post)

**Comment by sftim** [comment]
@nate-double-u thanks for taking that time to check!

**Comment by sftim** [comment]
```
curl -X DELETE \
localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{
"kind":"DeleteOptions",
"apiVersion":"v1",
"propagationPolicy":"Background"
}' \
  -H "Content-Type: application/json"
```

The problem here is wrapping I think.

```bash
curl -X DELETE \
  http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
  -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
  -H "Content-Type: application/json"
```
ought to work fine

**Comment by nate-double-u** [comment]
> ```shell
> curl -X DELETE \
>   http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
>   -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
>   -H "Content-Type: application/json"
> ```
> 
> ought to work fine

Confirmed, this works. Thanks!

File: content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "test"
+  },
+  "spec": {
+    "finalizers": null,
+  }
+}
+EOF
+```
```
Comment (nate-double-u) [inline content/en/blog/_posts/2021-04-28-using-finalizers-to-control-deletion.md:262]: Please see https://github.com/kubernetes/website/pull/26997#issuecomment-829063409

**Comment by aaron-kasten** [comment]
> ```shell
> ```shell
> curl -X DELETE \
>   http://localhost:8080/api/v1/namespaces/default/configmaps/mymap-parent \
>   -d '{ "kind":"DeleteOptions", "apiVersion":"v1", "propagationPolicy":"Background" }' \
>   -H "Content-Type: application/json"
> ```
> 
> 
> ought to work fine
> 
> Confirmed, this works. Thanks!
> ```

@nate-double-u is there action I should take on this?  This sounds like a layout issue?

**Comment by nate-double-u** [comment]
> @nate-double-u is there action I should take on this? This sounds like a layout issue?

I think that if folks are copying and pasting snippets from the article, then @sftim's layout works better (though it may be less readable as several bits of info are put onto the same line).

I think I'd err on the side of things working for the most readers when copied from the article, so I'd suggest updating the layout. That said, this may be an issue with my setup (using zsh instead of bash), so I'd defer the decision to you others who might have an opinion (does it work as is for other users?)

**Comment by aaron-kasten** [comment]
> > @nate-double-u is there action I should take on this? This sounds like a layout issue?
> 
> I think that if folks are copying and pasting snippets from the article, then @sftim's layout works better (though it may be less readable as several bits of info are put onto the same line).
> 
> I think I'd err on the side of things working for the most readers when copied from the article, so I'd suggest updating the layout. That said, this may be an issue with my setup (using zsh instead of bash), so I'd defer the decision to you and others who might have an opinion (does it work as is for other users?)

@nate-double-u agreed - I think a least-common-denominator approach is best, which is represented by keeping everything on one line.  The JSON is not so large that it makes the example confusing.

**Comment by aaron-kasten** [comment]
@sftim @nate-double-u changes are in ... all set.

**Comment by nate-double-u** [comment]
new date deploy preview: [Using Finalizers to Control Deletion](https://deploy-preview-26997--kubernetes-io-master-staging.netlify.app/blog/2021/04/30/using-finalizers-to-control-deletion/)

File: content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+  "metadata": {
+    "name": "test"
+  },
+  "spec": {
+    "finalizers": null,
+  }
+}
+EOF
+```
+
```
Comment (nate-double-u) [inline content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:260]: I'm still having trouble getting this snippet to work. I'm not sure if the setup is intended to be the same as the examples above, but I always get the following:

```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type Namespace): couldn't get version/kind; json parse error: invalid character '}' looking for beginning of object key string ({\n  \"kind\": \"Namespace\",\n  \"ap ...)",
  "reason": "BadRequest",
  "code": 400
}% 
```

File: content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+## The basic `delete`
+
+Kubernetes has several different commands you can use that allow you to create, read, update, and delete objects. For the purpose of this blog post, we’ll focus on four `kubectl` commands: `create`, `get`, `patch`, and `delete`. 
+
+Here are examples of the basic `kubectl delete` command:
+
+```
+$ kubectl create configmap mymap
+configmap/mymap created
```
Comment (kbhawkey) [inline content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:29]: minor nit: You could remove the shell prompt ($)

Comment (aaron-kasten) [reply content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:29]: I removed the command prompts.

File: content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md (hunk 12 above)
Comment (aaron-kasten) [reply content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md]: there was a dangling `,` - removed it.  Example should work now.

Comment (aaron-kasten) [reply content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md]: @nate-double-u done.

File: content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md (hunk 13 above)
Comment (aaron-kasten) [reply content/en/blog/_posts/2021-04-30-using-finalizers-to-control-deletion.md:29]: @kbhawkey done.

**Review by nate-double-u — COMMENTED** [review]
I'm still not sure how to get k8s into a state described in the _Forcing a Deletion of a Namespace_ section, but when I create a new test namespace (`kubectl create namespace test`) first, the snippet doesn't throw the `BadRequest` error.

Thanks for this @aaron-kasten, it's an interesting article!

LGTM

**Comment by aaron-kasten** [comment]
@sftim @kbhawkey unsure of what the next steps are.  The PR still needs a `lgtm` from a reviewer.

**Comment by aaron-kasten** [comment]
@sftim is this ready?

**Comment by sftim** [comment]
I've set a publication date for Friday
/label tide/merge-method-squash

**Pull request merged**